### PR TITLE
feat: E2E test using Playwright

### PR DESCRIPTION
**Pull request by gee05053**
This PR resolves https://github.com/lablup/backend.ai-webui/issues/2081

Implement e2e test using playwright
[Loop Doc](https://lablup.sharepoint.com/:fl:/s/fb28b4ec-f199-4e97-a2b6-c73a1fcdd1ff/ETnciCtsG5ZKmBN4dxi2_2gBAvbgtYNmqRoklJoqKSIylw?e=obSCqL&nav=cz0lMkZzaXRlcyUyRmZiMjhiNGVjLWYxOTktNGU5Ny1hMmI2LWM3M2ExZmNkZDFmZiZkPWIlMjFlS3Z6RE5UdHUwbWxLRVJwTEhPeWtibHFzYU12SDlkS2xLeTZHa29tcVUyUTBZcTdyUFJDVEpZdkxqVktWU2h3JmY9MDFXUkVWVkRKWjNTRUNXM0EzU1pGSlFFM1lPNE1MTjczSSZjPSUyRiZhPUxvb3BBcHAmcD0lNDBmbHVpZHglMkZsb29wLXBhZ2UtY29udGFpbmVyJng9JTdCJTIydyUyMiUzQSUyMlQwUlRVSHhzWVdKc2RYQXVjMmhoY21Wd2IybHVkQzVqYjIxOFlpRmxTM1o2UkU1VWRIVXdiV3hMUlZKd1RFaFBlV3RpYkhGellVMTJTRGxrUzJ4TGVUWkhhMjl0Y1ZVeVVUQlpjVGR5VUZKRFZFcFpka3hxVmt0V1UyaDNmREF4VjFKRlZsWkVTVmd5VEVaSVRVNWFTelpPUkVsTVUwVk1XRk5DUmtKTVFqVSUzRCUyMiUyQyUyMmklMjIlM0ElMjI3ZjMyMTZhMy1hZWJjLTRmZGEtYjkxMi1lYWRjNGIxMzFhYWYlMjIlN0Q%3D)

You must make .env file at playwright directory(webui/playwright/.env) and copy and paste below config
```
SUPER_ADMIN_EMAIL=admin@lablup.com
SUPER_ADMIN_PASSWORD=wJalrXUt
ADMIN_EMAIL=domain-admin@lablup.com
ADMIN_PASSWORD=cWbsM_vB
USER_EMAIL=user@lablup.com
USER_PASSWORD=C8qnIo29
USER2_EMAIL=user2@lablup.com
USER2_PASSWORD=P7oxTDdz
INCORRECT_EMAIL=test7@lablup.com
INCORRECT_PASSWORD=dasdnsajdnk
PAGE_URL=http://localhost:9081
ENDPOINT=http://localhost:8090
CREATE_VFOLDER_GROUP=test_peoject
NOT_CREATE_VFOLDER_GROUP=default
DELETE_VFOLDER=vMyFkaomKN9vkTORaywaleDayTjYGXwDyB4LPxUiLFvVAWCW2iFjEtJzzXZ8gNK3
NEW_SESSION_VERSION=Ubuntu 20.04 aarch64
NEW_SESSION_MOUNTED_VFOLDER=test
DELETE_SESSION=vNkj
CREATE_USER_NAME=test9
CREATE_USER_EMAIL=test9@lablup.com
CREATE_USER_PASSWORD=test123!
DELETE_USER_EMAIL=test9@lablup.com
```
How to run
1. Follow [extension guide](https://playwright.dev/docs/getting-started-vscode#running-tests) to install extension
2. Pull this branch
3. Open test explorer at left side bar as in the picture below
스크린샷 2023-12-26 오후 6 03 33
4. Click to play button at test explorer or beside of code lines to run test 
스크린샷 2023-12-26 오후 6 06 55
5. You can check the testing process through browser and test result on terminal
스크린샷 2023-12-26 오후 6 13 36
스크린샷 2023-12-26 오후 6 13 55

**Checklist:** (if applicable)

- [x] Mention to the original issue
- [x] Documentation
- [ ] Minium required manager version
- [ ] Specific setting for review (eg., KB link, endpoint or how to setup)
- [ ] Minimum requirements to check during review
- [ ] Test case(s) to demonstrate the difference of before/after

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,8 +28,7 @@
     "docs": "npx jsdoc -c ./jsdoc.json",
     "update": "npx tsc",
     "postinstall": "npx concurrently -c \"auto\" --names \"react-dependencies\" \"cd react && npm install\"",
-    "patch:lit-translate": "unzip -o ./scripts/lit-translate.zip -d ./node_modules/lit-translate && cp -f ./scripts/lit-translate-index.js ./node_modules/lit-translate/index.js",
-    "prepare": "husky install"
```
Comment (yomybaby) [inline package.json]: You might have deleted this line by mistake. Please check carefully before pushing changes and requesting a review. @gee05053

**Review by yomybaby — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (gee05053) [reply package.json]: @yomybaby
Sorry.. I will check one more time and request review again

**Comment by yomybaby** [comment]
@gee05053 This PR doesn't have any description of how to run the test.
Please add it with related links.

File: .prettierignore [hunk 2]
```diff
@@ -2,5 +2,6 @@
 src/lib/
 !/src
 !/react
+!/playwright
```
Comment (yomybaby) [inline .prettierignore:5]: Is there any reason to exclude `/playwright`?

Comment (gee05053) [reply .prettierignore:5]: @yomybaby
I tried to prettier test without this line.
But prettier was not applied.
I will check one more time.

Comment (gee05053) [reply .prettierignore:5]: @yomybaby 
```!/playwright``` must exclude to apply prettier.
Because first line is ```/*```

**Comment by yomybaby** [comment]
@gee05053 
When I run `npx playwright install` I got errors:
```
Error: test file "e2eTest.spec.ts" should not import test file "pages/CredentialPage.spec.ts"

   at e2eTest.spec.ts:1

> 1 | import { createUser, deleteUser } from './pages/CredentialPage.spec';
    | ^
  2 | import { createVfolder, deleteVfolder } from './pages/DataStoragePage.spec';
  3 | import { login } from './pages/LoginPage.spec';
  4 | import { createSession, deleteSession } from './pages/SessionPage.spec';

Error: test file "e2eTest.spec.ts" should not import test file "pages/DataStoragePage.spec.ts"

   at e2eTest.spec.ts:1

> 1 | import { createUser, deleteUser } from './pages/CredentialPage.spec';
    | ^
  2 | import { createVfolder, deleteVfolder } from './pages/DataStoragePage.spec';
  3 | import { login } from './pages/LoginPage.spec';
  4 | import { createSession, deleteSession } from './pages/SessionPage.spec';

Error: test file "e2eTest.spec.ts" should not import test file "pages/LoginPage.spec.ts"

   at e2eTest.spec.ts:1

> 1 | import { createUser, deleteUser } from './pages/CredentialPage.spec';
    | ^
  2 | import { createVfolder, deleteVfolder } from './pages/DataStoragePage.spec';
  3 | import { login } from './pages/LoginPage.spec';
  4 | import { createSession, deleteSession } from './pages/SessionPage.spec';

Error: test file "e2eTest.spec.ts" should not import test file "pages/SessionPage.spec.ts"

   at e2eTest.spec.ts:1

> 1 | import { createUser, deleteUser } from './pages/CredentialPage.spec';
    | ^
  2 | import { createVfolder, deleteVfolder } from './pages/DataStoragePage.spec';
  3 | import { login } from './pages/LoginPage.spec';
  4 | import { createSession, deleteSession } from './pages/SessionPage.spec';




To open last HTML report run:

  npx playwright show-report
```

File: playwright/e2eTest.spec.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { createUser, deleteUser } from './pages/CredentialPage.spec';
+import { createVfolder, deleteVfolder } from './pages/DataStoragePage.spec';
+import { login } from './pages/LoginPage.spec';
+import { createSession, deleteSession } from './pages/SessionPage.spec';
+import { test } from '@playwright/test';
+
+test('E2E test', async ({ page }) => {
```
Comment (yomybaby) [inline playwright/e2eTest.spec.ts]: Currently, one test runs all tests. This is not a good pattern. Please separate them. You can google  e2e test examples that shows how to organize and separate tests.

**Review by yomybaby — CHANGES_REQUESTED** [review]

File: playwright/e2eTest.spec.ts (hunk 3 above)
Comment (yomybaby) [reply playwright/e2eTest.spec.ts]: https://playwright.dev/docs/writing-tests#using-test-hooks

File: .eslintignore [hunk 4]
```diff
@@ -15,4 +15,4 @@ src/lib/backend.ai-client-node.ts
 src/lib/backend.ai-client-node-to-es6.js
 src/lib/backend.ai-client-node-test-native-es6.js
 src/lib/Chart.min.js
-src/lib/json_to_csv.js
+src/lib/json_to_csv.js
```
Comment (yomybaby) [inline .eslintignore]: Is there any reason to delete? If not, please restore.

File: src/components/backend-ai-session-list.ts [hunk 5]
```diff
@@ -2970,7 +2970,7 @@ ${rowData.item[this.sessionNameField]}</pre
             ? html`
                 <mwc-icon-button
                   class="fg red controls-running"
-                  id="${rowData.index + '-power'}"
+                  id="${rowData.item.name + '-power'}"
```
Comment (yomybaby) [inline src/components/backend-ai-session-list.ts]: For the test, use `data-test` instead of `id` if available

**Review by yomybaby — CHANGES_REQUESTED** [review]

File: .eslintignore (hunk 4 above)
Comment (gee05053) [reply .eslintignore]: @yomybaby Deleted code and new code is same. But I don't know why file is shown that was changed

File: playwright/pages/CredentialPage.spec.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
+import { test, expect } from '@playwright/test';
+import dotenv from 'dotenv';
+import path from 'path';
+
+dotenv.config();
+dotenv.config({ path: path.resolve(__dirname, '..', '.env') });
+
+test.describe('Create account', () => {
+  test.beforeEach(async ({ page }) => {
+    await page.goto('http://localhost:9081');
```
Comment (yomybaby) [inline playwright/pages/CredentialPage.spec.ts]: Let's manage this in `.env`

File: playwright/pages/LoginPage.spec.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      .fill(process.env.ENDPOINT as string);
+    await page.locator('#login-button').click();
+    await page.waitForResponse(async (response) => {
+      return (
+        response.url() === 'http://localhost:8090/server/login' &&
+        response.status() === 200 &&
+        (await response.json()).authenticated
+      );
+    });
+    await expect(page.locator('#app-body')).toBeVisible();
```
Comment (yomybaby) [inline playwright/pages/LoginPage.spec.ts]: Please update this part because React Root PR #2121 changed the main layout.

**Review by yomybaby — CHANGES_REQUESTED** [review]

File: playwright/pages/DataStoragePage.spec.ts [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    await page.locator('#add-folder-name').click();
+    await page.locator('#add-folder-name label').fill(randomVfolderName);
+    await page.locator('#add-button').click();
+    await page.waitForResponse(
+      (response) =>
+        response.url() ===
+          `${process.env.ENDPOINT}/func/folders?group_id=2de2b969-1d04-48a6-af16-0bc8adb3c831` &&
+        response.status() === 200,
+      { timeout: 0 },
+    );
```
Comment (yomybaby) [inline playwright/pages/DataStoragePage.spec.ts]: Please change this line to not use the hard-coded`group_id` (Teams [thread](https://teams.microsoft.com/l/message/19:14c484402d874dafb15806d093b95a82@thread.skype/1708061257328?tenantId=13c6a44d-9b52-4b9e-aa34-0513ee7131f2&groupId=74ae2c4d-ec4d-4fdf-b2c2-f5041d1e8631&parentMessageId=1703039554351&teamName=devops&channelName=Frontend&createdTime=1708061257328))

**Review by yomybaby — CHANGES_REQUESTED** [review]

**Comment by yomybaby** [comment]
There have been many changes in the UI after this PR was created. Therefore, I will close this PR without merging and continue working from https://github.com/lablup/backend.ai-webui/pull/2647 . @gee05053  Thank you for your contribution.

**Pull request closed without merging**